ziti/ziti.go

```diff
@@ -295,30 +293,52 @@ func (context *contextImpl) Authenticate() error {
 		return errors.Errorf("SdkInfo is no longer a map[string]interface{}. Cannot request configTypes!")
 	}
 	var err error
-	context.apiSession, err = context.ctrlClt.Login(info, context.config.ConfigTypes)
-	return err
+	if context.apiSession, err = context.ctrlClt.Login(info, context.config.ConfigTypes); err != nil {
+		return err
+	}
+
+	var doOnceErr error
+	context.firstAuthOnce.Do(func() {
+		go context.runSessionRefresh()
+
+		metricsTags := map[string]string{
+			"srcId": context.apiSession.Identity.Id,
+		}
+		
+		context.metrics = metrics.NewRegistry(context.apiSession.Identity.Name, metricsTags)
+
+		// get services
+		if services, err := context.getServices(); err != nil {
+			doOnceErr = err
+		} else {
+			context.processServiceUpdates(services)
+		}
+	})
+
+	return doOnceErr
 }
 
 func (context *contextImpl) Dial(serviceName string) (edge.ServiceConn, error) {
 	if err := context.initialize(); err != nil {
 		return nil, errors.Errorf("failed to initialize context: (%v)", err)
 	}
-	id, ok := context.getServiceId(serviceName)
+	serviceId, ok := context.getServiceId(serviceName)
 	if !ok {
 		return nil, errors.Errorf("service '%s' not found", serviceName)
 	}
 
 	var conn edge.ServiceConn
 	var err error
 	for attempt := 0; attempt < 2; attempt++ {
-		ns, err := context.GetSession(id)
+		session, err := context.GetSession(serviceId)
 		if err != nil {
 			return nil, err
 		}
-		conn, err = context.dialSession(serviceName, ns)
-		if err != nil && attempt == 0 {
+		pfxlog.Logger().Infof("connecting via session id [%s] token [%s]", session.Id, session.Token)
+		conn, err = context.dialSession(serviceName, session)
+		if err != nil {
 			if strings.Contains(err.Error(), "closed") {
```

Comment:
this match seems very broad

Reply:
I don't know what it is for or why it is there. IIRC I set a breakpoint on it and it was never triggered.